src/framework/mlt_animation.c

```diff
@@ -973,3 +963,289 @@ void mlt_animation_clear_string(mlt_animation self)
+ *
+ * If fewer than 2 points are available, then duplicate the first and/or last points as necessary to
+ * meet the requirement for 4 points.
+ */
+
+static inline double catmull_rom_interpolate(
+    double x0, double y0, double x1, double y1, double x2, double y2, double x3, double y3, double t)
+{
+    // Correct first and last values.
+    // If points are duplicated (e.g. for the first and last segments) assume the duplicated point
+    // is far away to create a horizontal segment.
+    if (x0 == x1) {
+        x0 -= 10000;
+    }
+    if (x3 == x2) {
+        x3 += 10000;
+    }
+    // Alpha:
+    //  0.0 for the uniform spline
+    //  0.5 for the centripetal spline (no cusps)
+    //  1.0 for the chordal spline.
+    const double alpha = 0.50;
+    // Tension: 0-1.0. Could be made user configurable in the future.
+    //  A value of 1 results in a horizontal tangent at x1,y1 and x2,y2 (slope of 0).
+    //  A value of 0 results in the most natural slope at x1,y1 and x2,y2
+    const double tension = 0;
+    // m1 and m2 are the slope of the tangent line at points x1,y1 and x2,y2 respectively.
+    // In the case of peaks (direction change), make the tangent slope 0 so that the slope change
+    // occurs on the point that the user expected it. However, for the case that direction is not
+    // changing through the point, calculate a natural slope for m1, m2 between adjacent points.
+    double m1 = 0;
+    double m2 = 0;
+    double t12 = pow(distance(x1, y1, x2, y2), alpha);
+    if ((y1 < y0 && y1 > y2) || (y1 > y0 && y1 < y2)) {
+        double t01 = pow(distance(x0, y0, x1, y1), alpha);
+        m1 = (1.0 - tension) * (y2 - y1 + t12 * ((y1 - y0) / t01 - (y2 - y0) / (t01 + t12)));
+    }
+    if ((y2 < y1 && y2 > y3) || (y2 > y1 && y2 < y3)) {
+        double t23 = pow(distance(x2, y2, x3, y3), alpha);
+        m2 = (1.0 - tension) * (y2 - y1 + t12 * ((y3 - y2) / t23 - (y3 - y1) / (t12 + t23)));
+    }
+    double a = 2.0 * (y1 - y2) + m1 + m2;
+    double b = -3.0 * (y1 - y2) - m1 - m1 - m2;
+    double c = m1;
+    double d = y1;
+    return a * t * t * t + b * t * t + c * t + d;
+}
+
+/** Interpolate a new animation item given a set of other items.
+ *
+ * \private \memberof mlt_animation_s
+ *
+ * \param item an unpopulated animation item to be interpolated.
+ *  The frame and keyframe_type fields must already be set. The value for "frame" is the postion
+ *  at which the value will be interpolated. The value for "keyframe_type" determines which
+ *  interpolation will be used.
+ * \param p a sequential array of 4 animation items. The frame value for item must lie between the
+    frame values for p[1] and p[2].
+ * \param fps the frame rate, which may be needed for converting a time string to frame units
+ * \param locale the locale, which may be needed for converting a string to a real number
+  * \return true if there was an error
+ */
+
+static int interpolate_item(mlt_animation_item item,
```

Comment:
Why did I move this function? The centripetal method requires the position of all 4 points for non-uniform points. So I need the position of each property that is being used to interpolate. Instead of passing the positions to the property interpolation function, I just used the animation items (which already have the frame position included). I could move it back and add more parameters if we prefer that - it would need a new name to not break ABI.